### PR TITLE
fix(mariadb): support 3.5.3 CommonJS bundles

### PR DESCRIPTION
**Pull request by crysmags**
### What does this PR do?

Restores MariaDB 3.5.3+ CommonJS tracing while preserving the ESM instrumentation introduced by #9734.

MariaDB 3.5.3 restored `require()` through generated single-file bundles. Those entry points no longer load the Query, Execute, Pool, or Cluster source modules used by the existing instrumentation, so this PR adds a CommonJS adapter around the stable exported factories and their runtime results.

The adapter covers:

- direct, pooled, acquired, pool-emitted, and clustered connections for both promise and callback APIs;
- query, execute, batch, import-file, prepared-statement, stream, and transaction operations;
- callback and promise continuation context, command errors, stream identity, and pool skip behavior;
- detached top-level `importFile` calls without treating their receiver as a connection;
- exact cluster-node metadata across broad selectors, round-robin selection, retries, and immediate remove/re-add cycles; and
- the original CommonJS namespace descriptors and named/default factory identity.

### Motivation

MariaDB 3.5.1 and 3.5.2 were ESM-only; #9734 restored tracing for those releases through ESM loader hooks. MariaDB 3.5.3 added CommonJS again, but its `dist/promise.cjs` and `dist/callback.cjs` files embed the connector internals in generated, minified bundles. Requiring MariaDB therefore never loads the original source files where the existing Query, Execute, Pool, and Cluster hooks attach.

Orchestrion cannot rewrite those original modules before bundling because bundling already happened when MariaDB published the package. Targeting the generated bundle instead would couple AST matches to generated and minified output that can change between MariaDB releases. This adapter consequently wraps the stable public factories and returned runtime methods. Cluster connections do not expose which node was selected, so cluster metadata attribution narrowly observes the retained `_selectPool` runtime method and carries the selected node through async-local context, including retries.

### Additional Notes

#9734 is now merged into `master`. Together, #9734 and this PR restore the intended support split:

- MariaDB 3.5.1-3.5.2: ESM instrumentation from #9734.
- MariaDB 3.5.3+: the same ESM path plus the CommonJS bundle adapter in this PR.

Validation:

- MariaDB 3.5.3 CommonJS promise/callback bundle suite: 40 passing.
- MariaDB 3.5.3 ESM integration matrix: 5 passing.
- MariaDB 3.4.5 CommonJS regression suite: 49 passing.
- `mariadb-bundle.js` coverage: 94.13% lines, 90.39% statements, and 95.76% functions.
- Changed-file lint: passing.
- `git diff --check`: passing.

The generated supported-version rows match MariaDB 3.5.3. Repository-wide supported-integration verification still reports unrelated generated-version drift already present on `master`; this PR does not include that unrelated churn.

### CODE REVIEW

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+
+        return typeof callback === 'function'
+          ? finishCh.runStores(ctx, callback, this, ...arguments)
+          : finishCh.runStores(ctx, noop, this)
+      }
+
+      if (typeof callback === 'function') {
+        arguments[arguments.length - 1] = shimmer.wrapCallback(callback, wrapper)
+      } else {
+        arguments.length = Math.max(arguments.length + 1, command.length)
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: Addressed in `6c4590c3d`: preserved each original method arity through forwarding wrappers and now replace occupied null or undefined callback slots. Regression coverage exercises connection, pool, import-file, and transaction callback shapes.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 2]
```diff
@@ -0,0 +1,935 @@
...
+      if (!startCh.hasSubscribers) return command.apply(this, arguments)
+
+      const owner = commandOwner ?? this
+      const ctx = { sql: preparedSql ?? normalizeSql(sql), conf: options }
+
+      startCommand(owner)
+
+      let result
+      try {
+        result = startCh.runStores(ctx, command, this, ...arguments)
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: Deferred from #9736: this is a confirmed pre-existing MariaDB-wide DBM gap in both the existing internal-hook path and the bundled path. Rewriting only this adapter would make CJS and ESM behavior inconsistent, while a generic first-argument rewrite would corrupt prepared, import-file, and transaction calls. It needs a dedicated cross-path, SQL-aware change with DBM regression coverage; this PR remains scoped to restoring pre-3.5.1 tracing parity.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 3]
```diff
@@ -0,0 +1,935 @@
...
+ *
+ * @param {object} pool
+ * @param {object} options
+ * @param {(connection: object, options: object) => object} wrapConnection
+ * @returns {void}
+ */
+function wrapPoolConnectionEvent (pool, options, wrapConnection) {
+  const onConnection = connection => wrapConnection(connection, options)
+
+  pool.prependListener('connection', onConnection)
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: Addressed in `6c4590c3d`: moved pool connection wrapping into the emit wrapper before EventEmitter dispatch, with promise and callback prepend-listener regression coverage.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+}
+
+/**
+ * Extracts SQL from MariaDB's string and object command forms.
+ *
+ * @param {unknown} command
+ * @returns {unknown}
+ */
+function normalizeSql (command) {
+  return command !== null && typeof command === 'object' ? command.sql : command
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Nit

```suggestion
  return command?.sql ?? command
```

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+
+        return typeof callback === 'function'
+          ? finishCh.runStores(ctx, callback, this, ...arguments)
+          : finishCh.runStores(ctx, noop, this)
+      }
+
+      if (typeof callback === 'function') {
+        arguments[arguments.length - 1] = shimmer.wrapCallback(callback, wrapper)
+      } else if (callbackIndex >= 0 && arguments.length > callbackIndex && arguments[callbackIndex] == null) {
+        arguments[callbackIndex] = wrapper()
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Are we safe that this always works? I would imagine removing these lines would work with the code below in all cases?

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+ * @param {string} sql
+ * @param {(options: object, sql: string, owner?: object, commandArity?: number) =>
+ *   (command: Function) => Function} createWrapper
+ * @returns {(transaction: Function) => Function}
+ */
+function createWrapTransaction (options, sql, createWrapper) {
+  return function wrapTransaction (transaction) {
+    const tracedTransaction = createWrapper(options, sql, undefined, transaction.length)(function () {
+      return skipCh.runStores({}, transaction, this, ...arguments)
+    })
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Mind splitting this into two commands? Calling the returned method is difficult to read for me :)

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 7]
```diff
@@ -0,0 +1,988 @@
...
+ *
+ * @param {object} options
+ * @param {unknown} sql
+ * @param {object} connection
+ * @returns {(execute: Function) => Function}
+ */
+function createWrapCallbackPreparedExecute (options, sql, connection) {
+  return function wrapExecute (execute) {
+    const executeWithCallback = createWrapCallbackCommand(options, sql, connection)(execute)
+    const executeWithPromise = createWrapPromiseCommand(options, sql, connection)(execute)
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Mind separating the invocations? :)

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 8]
```diff
@@ -0,0 +1,988 @@
...
+function createWrapPromiseImportFile (defaultOptions) {
+  return function wrapImportFile (importFile) {
+    return function (options) {
+      const wrapper = createWrapPromiseCommand(
+        normalizeOptions(defaultOptions, options),
+        IMPORT_FILE_RESOURCE,
+        undefined,
+        undefined,
+        false
+      )(importFile)
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Mind making two calls out of this? Directly invoked returned methods are always tricky for me to read.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 9]
```diff
@@ -0,0 +1,988 @@
...
+function createWrapCallbackImportFile (defaultOptions) {
+  return function wrapImportFile (importFile) {
+    return function (options) {
+      const wrapper = createWrapCallbackCommand(
+        normalizeOptions(defaultOptions, options),
+        IMPORT_FILE_RESOURCE,
+        undefined,
+        undefined,
+        false
+      )(importFile)
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Same as above

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 10]
```diff
@@ -0,0 +1,988 @@
...
+ */
+function wrapBundle (mariadb, factories) {
+  const descriptors = Object.getOwnPropertyDescriptors(mariadb)
+  const defaultDescriptors = Object.getOwnPropertyDescriptors(mariadb.default)
+
+  for (const [name, wrapper] of factories) {
+    const wrapped = shimmer.wrapFunction(mariadb[name], wrapper)
+    replaceDescriptorValue(descriptors[name], wrapped)
+    replaceDescriptorValue(defaultDescriptors[name], wrapped)
+  }
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: shimmer.wrap has options to replace potential getters. I think we could do that instead of getting the descriptors up front

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 11]
```diff
@@ -0,0 +1,988 @@
...
+
+  const defaultExport = Object.defineProperties(
+    Object.create(Object.getPrototypeOf(mariadb.default)),
+    defaultDescriptors
+  )
+  replaceDescriptorValue(descriptors.default, defaultExport)
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Do we have to replace the object itself?

File: packages/datadog-plugin-mariadb/test/bundle.spec.js [hunk 12]
```diff
@@ -0,0 +1,610 @@
...
+}
+
+/**
+ * Waits for work already queued with setImmediate.
+ *
+ * @returns {Promise<void>}
+ */
+function nextImmediate () {
+  return new Promise(resolve => setImmediate(resolve))
+}
```
Comment (BridgeAR) [inline packages/datadog-plugin-mariadb/test/bundle.spec.js]: We could just use `node:timers/promises` for that :)

**Review by BridgeAR — COMMENTED** [review]
Just some nits here. The code is looking very good overall.

Next to my manual review, the machine found one thing:

Could you add a test that checks the spans for

```js
const ping = connection.ping()
const commit = connection.commit()
await Promise.all([ping, commit])
```

That will likely produce the wrong span. It could be fixed with

```js
for (const method of ['changeUser', 'ping', 'reset']) {
  shimmer.wrap(connection, method, createTrackPromiseCommand())
}
```

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 4 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: much cleaner, updated!

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L103-L105

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 5 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: kept this branch. MariaDB reads the declared callback slot, so when the caller passes explicit null or undefined, appending after it leaves the tracing callback unused. I centralized this in setCommandCallback and kept the regression coverage.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L415-L430

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 6 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: sure thing, split the wrapper creation from the invocation.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L721-L727

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 7 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: split both callback and promise wrapper creation from invocation.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L850-L858

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 8 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: updated this to create the wrapper first, then invoke it.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L1359-L1370

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 9 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: same update on the callback path: wrapper creation and invocation are separate now.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L1381-L1392

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 10 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: much cleaner, switched the namespace getters to shimmer.wrap with replaceGetter.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L1402-L1409

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 11 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: no, removed the default-object replacement. The default export is wrapped in place and points at the same wrapped factories as the namespace.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-instrumentations/src/mariadb-bundle.js#L1404-L1414

File: packages/datadog-plugin-mariadb/test/bundle.spec.js (hunk 12 above)
Comment (crysmags) [reply packages/datadog-plugin-mariadb/test/bundle.spec.js]: updated to node:timers/promises 👍

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-plugin-mariadb/test/bundle.spec.js#L9

**Comment by crysmags** [comment]
@BridgeAR added the concurrent ping/commit regression for both promise and callback APIs. I included prepare/commit too since prepare queues work through the same transaction gate.

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-plugin-mariadb/test/bundle.spec.js#L206-L220

https://github.com/DataDog/dd-trace-js/blob/9aabccadbda9f6ffe4dc89933ecb1fdac3955c8f/packages/datadog-plugin-mariadb/test/bundle.spec.js#L631-L647

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 13]
```diff
@@ -0,0 +1,1453 @@
...
+
+/**
+ * Removes options for every cluster node matching a selector.
+ *
+ * @param {Map<string, ClusterNodeOptions>} optionsByIdentifier
+ * @param {string} pattern
+ * @returns {void}
+ */
+function removeClusterOptions (optionsByIdentifier, pattern) {
+  const regularExpression = new RegExp(pattern)
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:1204]: No change: MariaDB 3.5.3 itself constructs RegExp(pattern) before removing any nodes. Therefore remove('*') already throws upstream, while remove('primary') intentionally also matches identifiers such as primary2. The instrumentation bookkeeping mirrors those connector semantics and does not turn a successful connector call into an error.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 14]
```diff
@@ -0,0 +1,1453 @@
...
+ *
+ * @param {Function} defaultOptions
+ * @returns {(createConnection: Function) => Function}
+ */
+function createWrapPromiseConnectionFactory (defaultOptions) {
+  return function wrapCreateConnection (createConnection) {
+    return function (options) {
+      return createConnection.apply(this, arguments).then(connection => {
+        return wrapPromiseConnection(connection, normalizeOptions(defaultOptions, options))
+      })
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:1394]: No change in this PR: a compatibility probe against the MariaDB 3.4.5 source path and the 3.5.3 bundle confirms that both retain factory-time tracing metadata after changeUser. Changing only the bundled path would create a behavioral divergence, so live metadata refresh needs a separate cross-path change.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 15]
```diff
@@ -0,0 +1,1453 @@
...
+        selection,
+        runClusterGetConnection,
+        getConnection,
+        this,
+        arguments
+      )
+
+      return result.then(
+        connection => finishPromiseGetConnection(ctx, connection, selection.options ?? emptyOptions),
+        error => finishPromiseGetConnectionError(ctx, error)
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: Addressed in `fda0d76a2`: failed promise and callback cluster acquisitions now publish an acquire-error lifecycle, including both a selected node that cannot connect and a selector with no matching node.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 16]
```diff
@@ -0,0 +1,1453 @@
...
+}
+
+/**
+ * Observes the public acquire event forwarded by a bundled pool.
+ *
+ * @param {object} pool
+ * @returns {void}
+ */
+function observePoolAcquisitions (pool) {
+  pool.prependListener('acquire', () => recordPoolAcquisition(pool))
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: Addressed in `fda0d76a2`: bundled pools now restore the acquisition observer after removeAllListeners cleanup, preserving pool wait metrics and preventing SQL errors from being misclassified as acquisition failures. MariaDB binds the original emit function before factory-level instrumentation runs, so the later emit wrapper cannot observe acquire events; the regression test also covers Node's removeListener meta-event path.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 17]
```diff
@@ -0,0 +1,1531 @@
...
+  if (pending === undefined) return
+
+  while (pending.index < pending.acquisitions.length) {
+    const acquisition = pending.acquisitions[pending.index]
+    if (isPoolAcquisitionPending(acquisition)) return
+    pending.index++
+  }
+
+  pendingPoolAcquisitions.delete(pool)
+}
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js:212]: I think this is not always safe as is.

We could do:

```diff
diff --git a/packages/datadog-instrumentations/src/mariadb-bundle.js b/packages/datadog-instrumentations/src/mariadb-bundle.js
--- a/packages/datadog-instrumentations/src/mariadb-bundle.js
+++ b/packages/datadog-instrumentations/src/mariadb-bundle.js
@@ -28,6 +28,7 @@ const wrappedConnections = new WeakSet()
 const IMPORT_FILE_RESOURCE = 'IMPORT FILE'
 const noop = () => {}
+const POOL_ACQUISITION_COMPACTION_THRESHOLD = 1024
 const STATUS_IN_TRANSACTION = 1
 const transactionMethods = [
   ['beginTransaction', 'START TRANSACTION'],
@@ -57,7 +58,7 @@ const transactionMethods = [
  * @property {number} [start]
  */
 /** @typedef {import('node:async_hooks').AsyncLocalStorage<PoolAcquisition>} PoolAcquisitionStorage */
-/** @typedef {{ acquisitions: PoolAcquisition[], index: number }} PendingPoolAcquisitions */
+/** @typedef {{ acquisitions: Array<PoolAcquisition | undefined>, index: number }} PendingPoolAcquisitions */
 
 /** @type {ClusterSelectionStorage | undefined} */
 let clusterSelectionStorage
@@ -164,6 +165,26 @@ function isPoolAcquisitionPending (acquisition) {
   return !acquisition.acquired && !acquisition.finished && !acquisition.errorReported
 }
 
+/**
+ * Releases a consumed acquisition and periodically compacts its queue.
+ *
+ * @param {PendingPoolAcquisitions} pending
+ * @returns {void}
+ */
+function discardPoolAcquisition (pending) {
+  pending.acquisitions[pending.index++] = undefined
+
+  if (
+    pending.index < POOL_ACQUISITION_COMPACTION_THRESHOLD ||
+    pending.index * 2 < pending.acquisitions.length ||
+    pending.index === pending.acquisitions.length
+  ) {
+    return
+  }
+
+  pending.acquisitions = pending.acquisitions.slice(pending.index)
+  pending.index = 0
+}
+
 /**
  * Removes completed entries from the front of a pool's acquisition queue.
  *
@@ -178,9 +199,10 @@ function prunePoolAcquisitions (pool) {
   if (pending === undefined) return
 
   while (pending.index < pending.acquisitions.length) {
-    const acquisition = pending.acquisitions[pending.index]
+    const acquisition = /** @type {PoolAcquisition} */ (pending.acquisitions[pending.index])
     if (isPoolAcquisitionPending(acquisition)) return
-    pending.index++
+
+    discardPoolAcquisition(pending)
   }
 
   pendingPoolAcquisitions.delete(pool)
@@ -197,7 +219,9 @@ function takePoolAcquisition (pool) {
   if (pending === undefined) return
 
   while (pending.index < pending.acquisitions.length) {
-    const acquisition = pending.acquisitions[pending.index++]
+    const acquisition = /** @type {PoolAcquisition} */ (pending.acquisitions[pending.index])
+    discardPoolAcquisition(pending)
+
     if (isPoolAcquisitionPending(acquisition)) return acquisition
   }
 ```

**Review by BridgeAR — COMMENTED** [review]
This is looking very good!

I think it would be nice to have some streaming tests in addition similar to

```js
const sql = 'SELECT * FROM definitely_missing_stream_table'

await Promise.all([
  agent.assertFirstTraceSpan({
    resource: sql,
    meta: {
      [ERROR_TYPE]: ANY_STRING,
      [ERROR_MESSAGE]: ANY_STRING,
      [ERROR_STACK]: ANY_STRING,
    },
  }),
  assert.rejects(consumeStream(connection.queryStream(sql))),
])
```

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 17 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:212]: @BridgeAR very true, switched this to clear each consumed slot and compact once the prefix reaches 1,024 and at least half the array.

https://github.com/DataDog/dd-trace-js/blob/3b1cbc8cba0580ac319541cfbae3e5e4a3e8f2c7/packages/datadog-instrumentations/src/mariadb-bundle.js#L177-L225

Also added a real 1,025-query queued-pool test so the first compaction boundary and FIFO order are covered.

https://github.com/DataDog/dd-trace-js/blob/3b1cbc8cba0580ac319541cfbae3e5e4a3e8f2c7/packages/datadog-plugin-mariadb/test/bundle.spec.js#L431-L453

**Comment by crysmags** [comment]
@BridgeAR added the failing `queryStream` case too. It now pins the error type, message, and stack through the bundled stream lifecycle.

https://github.com/DataDog/dd-trace-js/blob/3b1cbc8cba0580ac319541cfbae3e5e4a3e8f2c7/packages/datadog-plugin-mariadb/test/bundle.spec.js#L186-L200

**Review by BridgeAR — APPROVED** [review]
LGTM! Thank you :)
This is quite a significant code change!

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 18]
```diff
@@ -0,0 +1,1595 @@
...
+ */
+function createWrapCallbackPreparedExecute (options, sql, connection) {
+  return function wrapExecute (execute) {
+    const wrapCallbackCommand = createWrapCallbackCommand(options, sql, connection)
+    const wrapPromiseCommand = createWrapPromiseCommand(options, sql, connection)
+    const executeWithCallback = wrapCallbackCommand(execute)
+    const executeWithPromise = wrapPromiseCommand(execute)
+
+    return function () {
+      const hasCallback = typeof arguments[1] === 'function' || typeof arguments[2] === 'function'
```
[the remark this answers is not in the document]
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:952]: No change: the premise that MariaDB 3.5.3 accepts statement.execute(callback) for a statement with no parameters is incorrect. Its Prepare.execute(values, opts, callback, stack) overload parsing never examines argument 0 for a callback, even when the SQL has zero placeholders. That call stores the function as values, leaves callback unset, and returns a Promise, so result.then is valid. The supported no-values callback form is statement.execute(undefined, callback) or statement.execute([], callback), and the existing argument-1 check already detects it; treating argument 0 as a callback would create a span whose callback never runs.

File: packages/datadog-plugin-mariadb/src/index.js [hunk 19]
```diff
@@ -16,6 +16,19 @@ class MariadbPlugin extends MySQLPlugin {
...
+   *
+   * @param {{ currentStore?: { span?: import('../../../..').Span }, poolWaitTime?: number }} ctx
+   * @returns {void}
+   */
+  finish (ctx) {
+    if (ctx.poolWaitTime !== undefined) {
+      ctx.currentStore?.span?.setTag(`${this.component}.pool.wait_time`, ctx.poolWaitTime)
+    }
+    super.finish(ctx)
+  }
```
Comment (BridgeAR) [inline packages/datadog-plugin-mariadb/src/index.js]: Are you certain this is needed? I believe with the recent changes, this should be obsolete.

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 20]
```diff
@@ -0,0 +1,1595 @@
...
+}
+
+/**
+ * Observes public acquire events forwarded by a bundled pool.
+ *
+ * @param {object} pool
+ * @returns {void}
+ */
+function observePoolAcquisitions (pool) {
+  const observer = () => recordPoolAcquisition(pool)
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js]: Just in case it is not active :)

```suggestion
  const observer = () => {
    if (connectionStartCh.hasSubscribers) recordPoolAcquisition(pool)
  }
```

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 21]
```diff
@@ -0,0 +1,1595 @@
...
+/**
+ * Wraps promise connections acquired from a bundled pool cluster.
+ *
+ * @param {ClusterSelectionStorage} selectionStorage
+ * @returns {(getConnection: Function) => Function}
+ */
+function createWrapPromiseClusterGetConnection (selectionStorage) {
+  return function wrapGetConnection (getConnection) {
+    return function () {
+      const ctx = {}
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js:1280]:
```suggestion
      if (!connectionStartCh.hasSubscribers) return getConnection.apply(this, arguments)
      const ctx = {}
```

That way we have a fast path :)

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 22]
```diff
@@ -0,0 +1,1595 @@
...
+/**
+ * Wraps callback connections acquired from a bundled pool cluster.
+ *
+ * @param {ClusterSelectionStorage} selectionStorage
+ * @returns {(getConnection: Function) => Function}
+ */
+function createWrapCallbackClusterGetConnection (selectionStorage) {
+  return function wrapGetConnection (getConnection) {
+    return function () {
+      const callback = arguments[arguments.length - 1]
```
Comment (BridgeAR) [inline packages/datadog-instrumentations/src/mariadb-bundle.js:1312]:
```suggestion
      if (!connectionStartCh.hasSubscribers) return getConnection.apply(this, arguments)
      const callback = arguments[arguments.length - 1]
```

**Review by BridgeAR — APPROVED** [review]
LGTM, just some minor things that we could polish and I am unsure if we really need the finish part

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 20 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js]: @BridgeAR added the subscriber guard in 4522f3729: https://github.com/DataDog/dd-trace-js/blob/4522f37299fc09c88598063b704302bc6b346992/packages/datadog-instrumentations/src/mariadb-bundle.js#L414-L418

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 21 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:1280]: @BridgeAR a direct return here would skip the only path that adds per-instance query wrappers and selected-node options to bundled cluster connections. A safe inactive path would still run selectionStorage and resolve through wrapPromiseConnection(connection, selection.options ?? emptyOptions), while skipping the connection channels and timing. I left this unchanged because the direct return would leave a retained connection uninstrumented if MariaDB tracing is re-enabled.

File: packages/datadog-instrumentations/src/mariadb-bundle.js (hunk 22 above)
Comment (crysmags) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:1312]: @BridgeAR same issue on the callback path. The safe inactive path would still replace the user callback, call wrapCallbackConnection(connection, selection.options ?? emptyOptions), and then invoke the original callback; only context publication and acquisition timing can be skipped. I left this unchanged because returning directly would leave that acquired connection unwrapped after re-enable.

File: packages/datadog-plugin-mariadb/src/index.js (hunk 19 above)
Comment (crysmags) [reply packages/datadog-plugin-mariadb/src/index.js]: @BridgeAR very true, the delayed-start change means poolWaitTime is already present before MySQLPlugin.bindStart() runs, so this override became obsolete. Removed in 4522f3729: https://github.com/DataDog/dd-trace-js/blob/4522f37299fc09c88598063b704302bc6b346992/packages/datadog-plugin-mysql/src/index.js#L58-L76

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 23]
```diff
@@ -0,0 +1,1597 @@
...
+/**
+ * Completes tracking for a pool command that has returned to the caller.
+ *
+ * @param {PoolAcquisition | undefined} acquisition
+ * @param {unknown} [error]
+ * @returns {void}
+ */
+function finishPoolCommandAcquisition (acquisition, error) {
+  if (acquisition === undefined) return
+  if (error && acquisition.measure) {
```
[the remark this answers is not in the document]
Comment (BridgeAR) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:353]: This does not seem to be anything introduced in this PR

File: packages/datadog-instrumentations/src/mariadb-bundle.js [hunk 24]
```diff
@@ -0,0 +1,1597 @@
...
+      const result = selectionStorage.run(
+        selection,
+        runClusterGetConnection,
+        getConnection,
+        this,
+        arguments
+      )
+
+      return result.then(
+        connection => finishPromiseGetConnection(ctx, connection, selection.options ?? emptyOptions),
```
[the remark this answers is not in the document]
Comment (BridgeAR) [reply packages/datadog-instrumentations/src/mariadb-bundle.js:1296]: This does not seem to be anything introduced in this PR

**Review by BridgeAR — APPROVED** [review]
